### PR TITLE
JSON serialization

### PR DESCRIPTION
**Pull request by CloverCho**
LF PR to be merged together with this PR: https://github.com/lf-lang/lingua-franca/pull/1125

### CODE REVIEW

File: src/core/federation.ts [hunk 1]
```diff
@@ -241,15 +241,19 @@ class RTIClient extends EventEmitter {
...
 
     /**
      * Establish the mapping between a federate port's action and its ID.
      * @param federatePortID The federate port's ID.
      * @param federatePort The federate port's action.
      */
-    public registerFederatePortAction<T extends Present>(federatePortID: number, federatePortAction: Action<Buffer>) {
-        Object.setPrototypeOf(federatePortAction, FederatePortAction.prototype);
+    public registerFederatePortAction<T extends Present>(federatePortID: number, federatePortAction: Action<T>) {
+        //Object.setPrototypeOf(federatePortAction, Action<T>.prototype);
```
Comment (hokeun) [inline src/core/federation.ts:256]: Please remove commented line.

File: src/core/federation.ts [hunk 2]
```diff
@@ -447,15 +453,18 @@ class RTIClient extends EventEmitter {
      * @param time The time of the message encoded as a 64 bit little endian
      * unsigned integer in a Buffer.
      */
-    public sendRTITimedMessage(data: Buffer, destFederateID: number, destPortID: number, time: Buffer) {
-        let msg = Buffer.alloc(data.length + 21);
+    public sendRTITimedMessage<T extends Present>(data: T, destFederateID: number, destPortID: number, time: Buffer) {
+        const value = Buffer.from(JSON.stringify(data), "utf-8");
+
```
Comment (hokeun) [inline src/core/federation.ts:458]: Please remove extra blank lines.

File: src/core/reactor.ts [hunk 3]
```diff
@@ -1784,7 +1784,7 @@ export class App extends Reactor {
      * @param destFederateID The federate ID that is the destination of this message.
```
Comment (hokeun) [inline src/core/reactor.ts:1784]: Can you update the comment on @param data? It's not a Buffer anymore.

File: src/core/reactor.ts [hunk 4]
```diff
@@ -1795,7 +1795,7 @@ export class App extends Reactor {
...
```
Comment (hokeun) [inline src/core/reactor.ts:1795]: Same for the comment on Buffer here.

**Review by hokeun — APPROVED** [review]
This looks good to me. @lhstrh Could you please take another look?

File: src/core/federation.ts [hunk 5]
```diff
@@ -241,15 +241,18 @@ class RTIClient extends EventEmitter {
 
     // The mapping between a federate port ID and the federate port action
     // scheduled upon reception of a message designated for that federate port.
-    private federatePortActionByID: Map<number, FederatePortAction> = new Map<number, FederatePortAction>();
+    
+    /**
+     * TODO: Specify Type in Action
```
Comment (lhstrh) [inline src/core/federation.ts:246]:
```suggestion
     * A mapping from port IDs to FederatePortAction instances. Unfortunately, the data type of the action has to be `any`,
     * meaning that the type checker cannot check whether uses of the action are type safe. 
     * In an alternative design, type information might be preserved. TODO(marten): Look into this.
```

**Review by lhstrh — APPROVED** [review]

**Pull request merged**